### DISCUSSION TITLE
what are the best practices for creating schema `id`s

### DISCUSSION THREAD

**Discussion by turbomam** [linkml/linkml#1748] [Q&A] [answered]
- How should they relate to the YAML file name?
- are there related best practices for using w3id.org

@sierra-moxon are you pleased with the w3id.org redirection for biolink? @sujaypatil96 just showed me something he considered broken or unpredictable redirection.

**Comment by turbomam** [comment]
see also
- https://github.com/microbiomedata/nmdc-schema/issues/1313

**Comment by turbomam** [comment]
Is it possible that the schema `id` redirection for many LinkML repos is 404?

**Reply by cmungall** [reply]
this should be fixed

**Reply by nlharris** [reply]
since it's in discussions, it may not come to anyone's notice. should i make an issue and point to this discussion for details?

**Comment by turbomam** [comment]
Is there a strategy for testing w3id .htaccess files outside of waiting for the w3id people to do their daily merge?

**Reply by cmungall** [reply]
Yes, see https://github.com/linkml/linkml/tree/main/tests/test_rewrite_rules

**Comment by cmungall** [comment, marked as answer]
For production schemas:

- the schema id SHOULD be a https URL
- if the schema id is a URL then it SHOULD be resolvable (this needs to be fixed in a number of repos)
- w3id is a good choice but this is ultimately up to the provider
- there SHOULD be resolvable URLs that are obtained from standard URLs using the id as the base, including:
    - `{id}.yaml` - source schema MUST be included
    - for others, see biolink config

Note that for schemas that are not in production, or are intended for demonstration or test purposes, there is no requirement for resolvability. Providers are encouraged not to submit a w3id.org request until the schema and namespace are stable and there is evidence of use by users external to any one project.
